test/test_dist_autograd.py

```diff
@@ -159,6 +177,306 @@ def test_rpc_complex_args(self):
+            t2 = torch.rand(6, 6, requires_grad=True)
+
+
+            with self.assertRaises(RuntimeError):
+                # This should throw an error.
+                dist.rpc('worker{}'.format(dst_rank), torch.matmul,
+                         args=(t1, t2))
+
+
+    @dist_init
+    def test_backward_simple(self):
+        dst_rank = self._next_rank(self.rank)
+        with dist_autograd.context() as context_id:
+            t1 = torch.rand((3, 3), requires_grad=True)
+            t2 = torch.rand((3, 3), requires_grad=True)
+            ret = dist.rpc('worker{}'.format(dst_rank), torch.add,
+                           args=(t1, t2))
+            self.assertEqual(torch.add(t1, t2), ret)
+            dist_autograd.backward([ret.sum()])
+
+            # Verify grads were accumulated appropriately.
+            grads = dist_autograd.get_gradients(context_id)
+            self.assertEqual(2, len(grads))
+            self.assertIn(t1, grads)
+            self.assertIn(t2, grads)
+            self.assertEqual(torch.ones(3, 3), grads[t1])
+            self.assertEqual(torch.ones(3, 3), grads[t2])
+
+    @dist_init
+    def test_backward_multiple_round_trips(self):
+        dst_rank = self.rank
+        with dist_autograd.context() as context_id:
+            t1 = torch.rand((3, 3), requires_grad=True)
+            t2 = torch.rand((3, 3))
+            t3 = torch.rand((3, 3), requires_grad=True)
+            t4 = torch.rand((3, 3))
+            t5 = torch.rand((3, 3), requires_grad=True)
+
+            # Multiple RPCs between different nodes.
+            val = dist.rpc('worker{}'.format(self._next_rank(dst_rank)), torch.add,
+                           args=(t1, t2))
+            val = dist.rpc('worker{}'.format(self._next_rank(dst_rank)), torch.mul,
+                           args=(t3, val))
+            s1 = dist.rpc('worker{}'.format(self._next_rank(dst_rank)), torch.stack,
+                          args=([t4, val],))
+            s2 = dist.rpc('worker{}'.format(self._next_rank(dst_rank)), torch.stack,
+                          args=([t5, val],))
+            val = dist.rpc('worker{}'.format(self._next_rank(dst_rank)), torch.bmm,
+                           args=(s1, s2))
+            val = dist.rpc('worker{}'.format(self._next_rank(dst_rank)), torch.matmul,
+                           args=(val, val))
+
+            # Now run backwards.
+            dist_autograd.backward([val.sum()])
+
+            # Verify grads were accumulated appropriately.
+            grads = dist_autograd.get_gradients(context_id)
+            self.assertEqual(3, len(grads))
+            self.assertIn(t1, grads)
+            self.assertNotEqual(torch.zeros_like(t1), grads[t1])
+            self.assertNotIn(t2, grads)
+            self.assertIn(t3, grads)
+            self.assertNotEqual(torch.zeros_like(t3), grads[t3])
+            self.assertNotIn(t4, grads)
+            self.assertIn(t5, grads)
+            self.assertNotEqual(torch.zeros_like(t5), grads[t5])
```

Comment:
would you please get the grads locally and verify dist grads are exactly equal to them? feel like just verify them NotEqual(torch.zeros_like(t5), ...) is not enough to show grads results are expected

same for other unit tests

Reply:
Seconding this comment! (one of those cases where a "loopback" implementation of the rpc interface would be helpful. Don't actually exercise the messaging, just implement it all locally!)